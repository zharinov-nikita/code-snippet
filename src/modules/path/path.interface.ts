export interface InterfacePath<Options, Path> {
  generate: (options: Options) => Path
}
