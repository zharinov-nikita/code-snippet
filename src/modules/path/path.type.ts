import { TypeConfig, TypeGeneratePath } from '../../types'

export interface TypeGenerateOptions {
  pathToSnippet: string
  config: TypeConfig
  name: string
  path: string
}

export type TypeGenerateSnippetFileNameOptions = Pick<TypeGenerateOptions, 'pathToSnippet' | 'name'>

export type TypeGenerateDefaultPath = Pick<TypeGeneratePath, 'pathSnippetFile' | 'pathWriteSnippetFile'>
export type TypeGenerateFlatPath = Pick<TypeGeneratePath, 'pathFlatSnippetFile' | 'pathFlatWriteSnippetFile'>
