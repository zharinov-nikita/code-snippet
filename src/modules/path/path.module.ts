import path from 'path'
import { enumSnippetName } from '../../enums'
import { TypeGeneratePath } from '../../types'
import { ModuleString } from '../string'
import { InterfacePath } from './path.interface'
import type {
    TypeGenerateDefaultPath,
    TypeGenerateFlatPath,
    TypeGenerateOptions,
    TypeGenerateSnippetFileNameOptions,
} from './path.type'

export class ModulePath
    implements InterfacePath<TypeGenerateOptions, TypeGeneratePath>
{
    private string: ModuleString

    constructor() {
        this.string = new ModuleString()
    }

    private generateSnippetFileName(
        options: TypeGenerateSnippetFileNameOptions
    ): string {
        return path
            .basename(options.pathToSnippet)
            .replaceAll(
                enumSnippetName.camelCase,
                this.string.toCamelCase(options.name)
            )
            .replaceAll(
                enumSnippetName.lowerKebabCase,
                this.string.toLowerKebabCase(options.name)
            )
            .replaceAll(
                enumSnippetName.lowerSnakeCase,
                this.string.toLowerSnakeCase(options.name)
            )
            .replaceAll(
                enumSnippetName.pascalCase,
                this.string.toPascalCase(options.name)
            )
            .replaceAll(
                enumSnippetName.upperKebabCase,
                this.string.toUpperKebabCase(options.name)
            )
            .replaceAll(
                enumSnippetName.upperSnakeCase,
                this.string.toUpperSnakeCase(options.name)
            )
    }

    private generateDefaultPath(
        options: TypeGenerateOptions
    ): TypeGenerateDefaultPath {
        const snippetFileName = this.generateSnippetFileName(options)
        const unformattedPath = path.join(options.config.pathToSnippet)
        const formattedPath = path.join(...[options.path, options.name])
        const dirname = path.dirname(options.pathToSnippet)
        const pathSnippetFile = path.join(
            ...[dirname.replaceAll(unformattedPath, formattedPath)]
        )
        const pathWriteSnippetFile = path.join(
            ...[pathSnippetFile, snippetFileName]
        )
        return { pathSnippetFile, pathWriteSnippetFile }
    }

    private generateFlatPath(
        options: TypeGenerateOptions
    ): TypeGenerateFlatPath {
        const snippetFileName = this.generateSnippetFileName(options)
        const pathFlatSnippetFile = path.join(...[options.path])
        const pathFlatWriteSnippetFile = path.join(
            ...[pathFlatSnippetFile, snippetFileName]
        )
        return { pathFlatSnippetFile, pathFlatWriteSnippetFile }
    }

    public generate(options: TypeGenerateOptions): TypeGeneratePath {
        const snippetFileName = this.generateSnippetFileName(options)
        const { pathSnippetFile, pathWriteSnippetFile } =
            this.generateDefaultPath(options)
        const { pathFlatSnippetFile, pathFlatWriteSnippetFile } =
            this.generateFlatPath(options)
        return {
            snippetFileName,
            pathSnippetFile,
            pathFlatSnippetFile,
            pathWriteSnippetFile,
            pathFlatWriteSnippetFile,
        }
    }
}
