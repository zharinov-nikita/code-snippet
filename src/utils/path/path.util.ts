import path from 'path'
import { enumSnippetName } from '../../enums'
import { TypeConfigItem, TypeGeneratePath } from '../../types'

interface TypeGenerateOptions {
  pathToSnippet: string
  config: TypeConfigItem
  name: string
  path: string
}

export class UtilPath {
  public generate(options: TypeGenerateOptions): TypeGeneratePath {
    const snippetFileName = path
      .basename(options.pathToSnippet)
      .replaceAll(enumSnippetName.lowerKebabCase, options.name)
    const pathSnippetFile = path.join(
      ...[
        path
          .dirname(options.pathToSnippet)
          .replaceAll(path.join(options.config.files), path.join(...[options.path, options.name])),
      ]
    )
    const pathWriteSnippetFile = path.join(
      ...[
        pathSnippetFile,
        path.basename(options.pathToSnippet).replaceAll(enumSnippetName.lowerKebabCase, options.name),
      ]
    )

    return { snippetFileName, pathSnippetFile, pathWriteSnippetFile }
  }
}
