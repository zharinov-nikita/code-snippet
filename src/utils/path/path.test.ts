import { describe, expect, test } from '@jest/globals'
import { UtilPath } from './path.util'

describe('util/path', () => {
  const util = new UtilPath()

  test('snapshot', () => {
    expect(util).toMatchSnapshot('util')
  })
})
