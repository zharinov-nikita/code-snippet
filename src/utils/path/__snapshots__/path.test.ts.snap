// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util/path snapshot: util 1`] = `UtilPath {}`;
